test/hotspot/jtreg/serviceability/jvmti/FollowReferences/FieldIndices/libFieldIndicesTest.cpp

```diff
@@ -0,0 +1,517 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include <string.h>
+#include <stdint.h>
+#include "jvmti.h"
+#include "jvmti_common.h"
+
+extern "C" {
+
+static jvmtiEnv* jvmti = nullptr;
+static bool test_failed = false;
+
+static void* allocate(JNIEnv* env, jlong size) {
+    unsigned char* result = nullptr;
+    check_jvmti_status(env,
+        jvmti->Allocate(size, &result),
+        "Allocate failed");
+    return result;
+}
+
+static void deallocate(JNIEnv* env, void* mem) {
+    check_jvmti_status(env,
+        jvmti->Deallocate((unsigned char*)mem),
+        "Deallocate failed");
+}
+
+// Converts JNI class name signature to simple name (in place).
+static void sig2name(char* str) {
+    size_t len = strlen(str);
+    if (len >=2 && str[0] == 'L' && str[len-1] == ';') {
+        len -=2;
+        memmove(str, str+1, len);
+        str[len] = '\0';
+    }
+    // Replace '/' with '.'.
+    for (char* pos = str; (pos = strchr(pos, '/')) != nullptr; ) {
+        *pos = '.';
+    }
+}
+
+static bool is_static_field(JNIEnv* env, jclass klass, jfieldID fid) {
+    enum {
+       ACC_STATIC        = 0x0008,
+    };
+
+    jint access_flags = 0;
+    check_jvmti_status(env,
+        jvmti->GetFieldModifiers(klass, fid, &access_flags),
+        "GetFieldModifiers failed");
+    return (access_flags & ACC_STATIC) != 0;
+}
+
+static void verify_int_field(JNIEnv* env, jclass klass, jfieldID fid) {
+    char* name = nullptr;
+    char* sig = nullptr;
+    check_jvmti_status(env,
+        jvmti->GetFieldName(klass, fid, &name, &sig, nullptr),
+        "GetFieldName failed");
+
+    if (strcmp(sig, "I") != 0) {
+        printf("ERROR: field '%s' is not int ('%s')\n", name, sig);
+        fflush(0);
+        fatal(env, "unexpected field type");
+    }
+
+    deallocate(env, name);
+    deallocate(env, sig);
+}
+
+
+/*
+Per jvmtiHeapReferenceInfoField spec:
```

Comment:
I think you should also include mention that this is for JVMTI_HEAP_REFERENCE_STATIC_FIELD.

Indentation should be fixed to match our C++ comment style, and the "bullets" from the spec should be included to make it clearer that these are two lists of steps.

Reply:
Fixed.